{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract transcript ID and gene name From GFF file\n",
    "\n",
    "This is a simple script to extract transcript IDs and gene names from genomic GFF file GRCh38_latest_genomic.gff (obtained from https://www.ncbi.nlm.nih.gov/projects/genome/guide/human/index.shtml) to a tab-separated files and bed file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# We want to get rid of the first 7 lines of the gff file \n",
    "with open(\"../data/GRCh38_latest_genomic.gff\") as f:\n",
    "    all_lines = f.readlines()\n",
    "with open(\"../data/GRCh38_latest_genomic_WithoutHeaderLines.tsv\",'w') as fw:\n",
    "    fw.writelines(all_lines[8:])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3695771, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>248956422.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:9606;Name=1;chromosome=1;g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:100287102,HGNC:HGNC:371...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:100287102,G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:100287102,Gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:100287102,Gen...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0           1           2        3            4  5  6  7  \\\n",
       "0  NC_000001.11      RefSeq      region      1.0  248956422.0  .  +  .   \n",
       "1  NC_000001.11  BestRefSeq  pseudogene  11874.0      14409.0  .  +  .   \n",
       "2  NC_000001.11  BestRefSeq  transcript  11874.0      14409.0  .  +  .   \n",
       "3  NC_000001.11  BestRefSeq        exon  11874.0      12227.0  .  +  .   \n",
       "4  NC_000001.11  BestRefSeq        exon  12613.0      12721.0  .  +  .   \n",
       "\n",
       "                                                   8  \n",
       "0  ID=id0;Dbxref=taxon:9606;Name=1;chromosome=1;g...  \n",
       "1  ID=gene0;Dbxref=GeneID:100287102,HGNC:HGNC:371...  \n",
       "2  ID=rna0;Parent=gene0;Dbxref=GeneID:100287102,G...  \n",
       "3  ID=id1;Parent=rna0;Dbxref=GeneID:100287102,Gen...  \n",
       "4  ID=id2;Parent=rna0;Dbxref=GeneID:100287102,Gen...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's read in all the transcript IDs\n",
    "alltranscripts = pd.read_csv(\"../data/GRCh38_latest_genomic_WithoutHeaderLines.tsv\",header=None,sep=\"\\t\")\n",
    "print alltranscripts.shape\n",
    "alltranscripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2097281, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:100287102,G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:100287102,Gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:100287102,Gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id3;Parent=rna0;Dbxref=GeneID:100287102,Gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14362.0</td>\n",
       "      <td>29370.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna1;Parent=gene1;Dbxref=GeneID:653635,Genb...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0           1           2        3        4  5  6  7  \\\n",
       "0  NC_000001.11  BestRefSeq  transcript  11874.0  14409.0  .  +  .   \n",
       "1  NC_000001.11  BestRefSeq        exon  11874.0  12227.0  .  +  .   \n",
       "2  NC_000001.11  BestRefSeq        exon  12613.0  12721.0  .  +  .   \n",
       "3  NC_000001.11  BestRefSeq        exon  13221.0  14409.0  .  +  .   \n",
       "4  NC_000001.11  BestRefSeq  transcript  14362.0  29370.0  .  -  .   \n",
       "\n",
       "                                                   8  \n",
       "0  ID=rna0;Parent=gene0;Dbxref=GeneID:100287102,G...  \n",
       "1  ID=id1;Parent=rna0;Dbxref=GeneID:100287102,Gen...  \n",
       "2  ID=id2;Parent=rna0;Dbxref=GeneID:100287102,Gen...  \n",
       "3  ID=id3;Parent=rna0;Dbxref=GeneID:100287102,Gen...  \n",
       "4  ID=rna1;Parent=gene1;Dbxref=GeneID:653635,Genb...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only grab rows that have both a gene name and a transcript ID\n",
    "transcripts_valid = alltranscripts[alltranscripts[8].str.contains(\"gene=\")&alltranscripts[8].str.contains(\"transcript_id=\")]\n",
    "print transcripts_valid.shape\n",
    "# Reset index \n",
    "transcripts_valid = transcripts_valid.reset_index(drop=True)\n",
    "transcripts_valid.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a small function to extract gene name and transcript ID for each row\n",
    "def extractGeneNameAndTranscriptID(row):\n",
    "    split_row = row.split(';')\n",
    "    genename = [i for i in split_row if \"gene=\" in i][0].split('=')[1]\n",
    "    transcriptID = [i for i in split_row if \"transcript_id=\" in i][0].split('=')[1]\n",
    "    return [genename,transcriptID]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use function to extract gene name and transcript IDs\n",
    "geneNamesTranscriptIDs = transcripts_valid[8].apply(extractGeneNameAndTranscriptID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>TranscriptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene TranscriptID\n",
       "0  DDX11L1  NR_046018.2\n",
       "1  DDX11L1  NR_046018.2\n",
       "2  DDX11L1  NR_046018.2\n",
       "3  DDX11L1  NR_046018.2\n",
       "4   WASH7P  NR_024540.1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert gene names and transcript IDs to a dataframe\n",
    "geneNamesTranscriptIDs_DF = pd.DataFrame(list(geneNamesTranscriptIDs.values))\n",
    "geneNamesTranscriptIDs_DF.columns = [\"Gene\",\"TranscriptID\"]\n",
    "geneNamesTranscriptIDs_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop duplicate values and write this data frame to a file\n",
    "geneNamesTranscriptIDs_DF[[\"TranscriptID\",\"Gene\"]].drop_duplicates().to_csv(\"../data/GeneNameTranscriptID_NCBI_RefSeq_hg38_FromGFFfile.tsv\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcriptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874.0</td>\n",
       "      <td>12227.0</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613.0</td>\n",
       "      <td>12721.0</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14362.0</td>\n",
       "      <td>29370.0</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chrm      RefSeq     feature    start      end strand     gene  \\\n",
       "0  NC_000001.11  BestRefSeq  transcript  11874.0  14409.0      +  DDX11L1   \n",
       "1  NC_000001.11  BestRefSeq        exon  11874.0  12227.0      +  DDX11L1   \n",
       "2  NC_000001.11  BestRefSeq        exon  12613.0  12721.0      +  DDX11L1   \n",
       "3  NC_000001.11  BestRefSeq        exon  13221.0  14409.0      +  DDX11L1   \n",
       "4  NC_000001.11  BestRefSeq  transcript  14362.0  29370.0      -   WASH7P   \n",
       "\n",
       "  transcriptID  \n",
       "0  NR_046018.2  \n",
       "1  NR_046018.2  \n",
       "2  NR_046018.2  \n",
       "3  NR_046018.2  \n",
       "4  NR_024540.1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# New data frame\n",
    "transcripts_valid_withNames = pd.concat([transcripts_valid[[0,1,2,3,4,6]],geneNamesTranscriptIDs_DF],axis=1,ignore_index=True)\n",
    "transcripts_valid_withNames.columns = [\"chrm\",\"RefSeq\",\"feature\",\"start\",\"end\",\"strand\",\"gene\",\"transcriptID\"]\n",
    "transcripts_valid_withNames.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2017619, 8)\n"
     ]
    }
   ],
   "source": [
    "# Keep only rows that have a valid chromsome that is not a scaffold. Check if NC\n",
    "transcripts_valid_withNames_validChr = transcripts_valid_withNames[transcripts_valid_withNames[\"chrm\"].str.contains(\"NC\")]\n",
    "print transcripts_valid_withNames_validChr.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NC_000024.10': 'chr24', 'NC_000006.12': 'chr6', 'NC_000011.10': 'chr11', 'NC_000023.11': 'chr23', 'NC_000013.11': 'chr13', 'NC_000003.12': 'chr3', 'NC_000009.12': 'chr9', 'NC_000022.11': 'chr22', 'NC_000001.11': 'chr1', 'NC_000004.12': 'chr4', 'NC_000018.10': 'chr18', 'NC_000016.10': 'chr16', 'NC_000015.10': 'chr15', 'NC_000008.11': 'chr8', 'NC_000002.12': 'chr2', 'NC_000007.14': 'chr7', 'NC_000010.11': 'chr10', 'NC_000014.9': 'chr14', 'NC_000012.12': 'chr12', 'NC_000019.10': 'chr19', 'NC_000017.11': 'chr17', 'NC_000005.10': 'chr5', 'NC_000020.11': 'chr20', 'NC_000021.9': 'chr21'}\n"
     ]
    }
   ],
   "source": [
    "# Dictionary to replace chromosome name \n",
    "chrms_to_replace = list(transcripts_valid_withNames_validChr[\"chrm\"].drop_duplicates().values)\n",
    "dict_chrms = {}\n",
    "for i in chrms_to_replace:\n",
    "    x = i.split('.')[0]\n",
    "    if x[len(x)-2]=='0':\n",
    "        dict_chrms[i]='chr'+x[len(x)-1]\n",
    "    else:\n",
    "        dict_chrms[i]='chr'+x[len(x)-2:]\n",
    "print dict_chrms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace chrm column with dictionary values\n",
    "transcripts_valid_withNames_validChr=transcripts_valid_withNames_validChr.replace({\"chrm\": dict_chrms})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcriptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrm      RefSeq     feature  start    end strand     gene transcriptID\n",
       "0  chr1  BestRefSeq  transcript  11874  14409      +  DDX11L1  NR_046018.2\n",
       "1  chr1  BestRefSeq        exon  11874  12227      +  DDX11L1  NR_046018.2\n",
       "2  chr1  BestRefSeq        exon  12613  12721      +  DDX11L1  NR_046018.2\n",
       "3  chr1  BestRefSeq        exon  13221  14409      +  DDX11L1  NR_046018.2\n",
       "4  chr1  BestRefSeq  transcript  14362  29370      -   WASH7P  NR_024540.1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Change start and end values to int\n",
    "transcripts_valid_withNames_validChr['start'] = transcripts_valid_withNames_validChr['start'].astype('int')\n",
    "transcripts_valid_withNames_validChr['end'] = transcripts_valid_withNames_validChr['end'].astype('int')\n",
    "transcripts_valid_withNames_validChr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort by chrm and start and end\n",
    "transcripts_valid_withNames_validChr_sorted = transcripts_valid_withNames_validChr.sort_values(by=[\"chrm\",\"start\",\"end\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write some of the columns to a file\n",
    "transcripts_valid_withNames_validChr_sorted[[\"chrm\",\"start\",\"end\",\"strand\",\"gene\",\"transcriptID\",\"RefSeq\",\"feature\"]].to_csv(\"../data/GRCh38_latest_genomic_ValidChroms_OnlyGenes.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1859728, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcriptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrm      RefSeq feature  start    end strand     gene transcriptID\n",
       "1   chr1  BestRefSeq    exon  11874  12227      +  DDX11L1  NR_046018.2\n",
       "2   chr1  BestRefSeq    exon  12613  12721      +  DDX11L1  NR_046018.2\n",
       "3   chr1  BestRefSeq    exon  13221  14409      +  DDX11L1  NR_046018.2\n",
       "15  chr1  BestRefSeq    exon  14362  14829      -   WASH7P  NR_024540.1\n",
       "14  chr1  BestRefSeq    exon  14970  15038      -   WASH7P  NR_024540.1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's get transcript IDs that exons \n",
    "transcripts_exons = transcripts_valid_withNames_validChr_sorted[transcripts_valid_withNames_validChr_sorted[\"feature\"]==\"exon\"]\n",
    "print transcripts_exons.shape\n",
    "transcripts_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157829, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcriptID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_000014.5</th>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000015.2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000016.5</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000017.3</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000018.3</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              chrm  RefSeq  feature  start  end  strand  gene\n",
       "transcriptID                                                 \n",
       "NM_000014.5     36      36       36     36   36      36    36\n",
       "NM_000015.2      2       2        2      2    2       2     2\n",
       "NM_000016.5     12      12       12     12   12      12    12\n",
       "NM_000017.3     10      10       10     10   10      10    10\n",
       "NM_000018.3     20      20       20     20   20      20    20"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by transcript ID to check how many exons are available for the transcript\n",
    "transcripts_exons_groupedById = transcripts_exons.groupby(by=[\"transcriptID\"]).agg('count')\n",
    "print transcripts_exons_groupedById.shape\n",
    "transcripts_exons_groupedById.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(153118, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>RefSeq</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcriptID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NM_000014.5</th>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000015.2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000016.5</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000017.3</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NM_000018.3</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              chrm  RefSeq  feature  start  end  strand  gene\n",
       "transcriptID                                                 \n",
       "NM_000014.5     36      36       36     36   36      36    36\n",
       "NM_000015.2      2       2        2      2    2       2     2\n",
       "NM_000016.5     12      12       12     12   12      12    12\n",
       "NM_000017.3     10      10       10     10   10      10    10\n",
       "NM_000018.3     20      20       20     20   20      20    20"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only grab transcript IDs of those that have greater than 1 exon\n",
    "transcripts_exons_groupedById_MoreThan1Exon = transcripts_exons_groupedById[transcripts_exons_groupedById[\"chrm\"]>1]\n",
    "print transcripts_exons_groupedById_MoreThan1Exon.shape\n",
    "transcripts_exons_groupedById_MoreThan1Exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "153118"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Grab the transcript IDs of transcripts with more than 1 exon\n",
    "transcriptIDs_MoreThan1Exon = list(set(list(transcripts_exons_groupedById_MoreThan1Exon.index.values)))\n",
    "len(transcriptIDs_MoreThan1Exon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(153118, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>TranscriptID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_001737835.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>NR_026818.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>LOC100996442</td>\n",
       "      <td>XR_001737582.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene    TranscriptID\n",
       "0        DDX11L1     NR_046018.2\n",
       "4         WASH7P     NR_024540.1\n",
       "18   MIR1302-2HG  XR_001737835.1\n",
       "24       FAM138A     NR_026818.1\n",
       "30  LOC100996442  XR_001737582.2"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get corresponding gene names\n",
    "transcriptIDsGeneName_MoreThan1Exon= geneNamesTranscriptIDs_DF[geneNamesTranscriptIDs_DF[\"TranscriptID\"].isin(transcriptIDs_MoreThan1Exon)]\n",
    "print transcriptIDsGeneName_MoreThan1Exon.drop_duplicates().shape\n",
    "transcriptIDsGeneName_MoreThan1Exon.drop_duplicates().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write data frame to a file\n",
    "transcriptIDsGeneName_MoreThan1Exon.drop_duplicates().to_csv(\"../data/GeneNameTranscriptID_NCBI_RefSeq_hg38_FromGFFfile_OnlyOnesWithIntrons.tsv\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
