{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extract introns and genome coordinates of genes from GFF file fata\n",
    "The purpose of this script is to extract intron coordinates and gene coordinates from the GFF file data\n",
    "We will look at features labeled exons and grab the end coordinates and start coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2017619, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2  3        4            5           6           7\n",
       "0  chr1  11874  12227  +  DDX11L1  NR_046018.2  BestRefSeq        exon\n",
       "1  chr1  11874  14409  +  DDX11L1  NR_046018.2  BestRefSeq  transcript\n",
       "2  chr1  12613  12721  +  DDX11L1  NR_046018.2  BestRefSeq        exon\n",
       "3  chr1  13221  14409  +  DDX11L1  NR_046018.2  BestRefSeq        exon\n",
       "4  chr1  14362  14829  -   WASH7P  NR_024540.1  BestRefSeq        exon"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_genes = pd.read_csv(\"../data/GRCh38_latest_genomic_ValidChroms_OnlyGenes.bed\",sep=\"\\t\",header=None)\n",
    "print data_genes.shape\n",
    "data_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157829, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>NR_106918.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>XR_001737835.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>NR_036051.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              4               5\n",
       "0       DDX11L1     NR_046018.2\n",
       "4        WASH7P     NR_024540.1\n",
       "11    MIR6859-1     NR_106918.1\n",
       "18  MIR1302-2HG  XR_001737835.1\n",
       "20    MIR1302-2     NR_036051.1"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get number of unique genes and transcript IDs \n",
    "genes_transcripts = data_genes.iloc[:,[4,5]]\n",
    "genes_transcripts_uniq = genes_transcripts.drop_duplicates()\n",
    "print genes_transcripts_uniq.shape\n",
    "genes_transcripts_uniq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set column names\n",
    "data_genes.columns = [\"chrm\",\"start\",\"end\",\"strand\",\"gene\",\"transcriptID\",\"RefSeqType\",\"feature\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1859728, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>RefSeqType</th>\n",
       "      <th>feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrm  start    end strand     gene transcriptID  RefSeqType feature\n",
       "0  chr1  11874  12227      +  DDX11L1  NR_046018.2  BestRefSeq    exon\n",
       "2  chr1  12613  12721      +  DDX11L1  NR_046018.2  BestRefSeq    exon\n",
       "3  chr1  13221  14409      +  DDX11L1  NR_046018.2  BestRefSeq    exon\n",
       "4  chr1  14362  14829      -   WASH7P  NR_024540.1  BestRefSeq    exon\n",
       "6  chr1  14970  15038      -   WASH7P  NR_024540.1  BestRefSeq    exon"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only grab exons \n",
    "data_exons = data_genes[data_genes[\"feature\"]==\"exon\"]\n",
    "print data_exons.shape\n",
    "data_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_exons.iloc[:,[0,1,2,5,3]].to_csv(\"../data/ExtractedExonCoordinatesFromGFFfile.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Let's get the exon-intron boundary coordinates from this exon data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    chrm  start    end strand         gene    transcriptID  RefSeqType feature\n",
      "0   chr1  11874  12227      +      DDX11L1     NR_046018.2  BestRefSeq    exon\n",
      "2   chr1  12613  12721      +      DDX11L1     NR_046018.2  BestRefSeq    exon\n",
      "3   chr1  13221  14409      +      DDX11L1     NR_046018.2  BestRefSeq    exon\n",
      "18  chr1  29926  30039      +  MIR1302-2HG  XR_001737835.1      Gnomon    exon\n",
      "21  chr1  30366  30503      +    MIR1302-2     NR_036051.1  BestRefSeq    exon\n",
      "   chrm  start    end strand    gene transcriptID  RefSeqType feature\n",
      "4  chr1  14362  14829      -  WASH7P  NR_024540.1  BestRefSeq    exon\n",
      "6  chr1  14970  15038      -  WASH7P  NR_024540.1  BestRefSeq    exon\n",
      "7  chr1  15796  15947      -  WASH7P  NR_024540.1  BestRefSeq    exon\n",
      "8  chr1  16607  16765      -  WASH7P  NR_024540.1  BestRefSeq    exon\n",
      "9  chr1  16858  17055      -  WASH7P  NR_024540.1  BestRefSeq    exon\n"
     ]
    }
   ],
   "source": [
    "# Split up exons by plus and minus strand\n",
    "data_exons_plus = data_exons[data_exons[\"strand\"]==\"+\"]\n",
    "print data_exons_plus.head()\n",
    "data_exons_minus = data_exons[data_exons[\"strand\"]==\"-\"]\n",
    "print data_exons_minus.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_exons_plus = data_exons_plus.reset_index(drop=True)\n",
    "data_exons_minus = data_exons_minus.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   chrm  start    end    transcriptID strand         gene\n",
      "0  chr1  12127  12327     NR_046018.2      +      DDX11L1\n",
      "1  chr1  12621  12821     NR_046018.2      +      DDX11L1\n",
      "2  chr1  14309  14509     NR_046018.2      +      DDX11L1\n",
      "3  chr1  29939  30139  XR_001737835.1      +  MIR1302-2HG\n",
      "4  chr1  30403  30603     NR_036051.1      +    MIR1302-2\n",
      "        chrm      start        end    transcriptID strand          gene\n",
      "919719  chr9  138044779  138044979     NR_121582.1      -  LOC101928786\n",
      "919720  chr9  138199833  138200033     NR_135133.1      -  LOC101928932\n",
      "919721  chr9  138203038  138203238     NR_135133.1      -  LOC101928932\n",
      "919722  chr9  138215881  138216081  XR_001746987.2      -  LOC107987145\n",
      "919723  chr9  138216445  138216645  XR_001746987.2      -  LOC107987145\n"
     ]
    }
   ],
   "source": [
    "# Get 5' exon-intron boundaries\n",
    "data_5p = pd.DataFrame({\"chrm\":pd.concat([data_exons_plus[\"chrm\"],data_exons_minus[\"chrm\"]]),\"start\":pd.concat([data_exons_plus[\"end\"]-100,data_exons_minus[\"start\"]-100]),\n",
    "                       \"end\":pd.concat([data_exons_plus[\"end\"]+100,data_exons_minus[\"start\"]+100]), \"strand\":pd.concat([data_exons_plus[\"strand\"],data_exons_minus[\"strand\"]]),\n",
    "                       \"transcriptID\":pd.concat([data_exons_plus[\"transcriptID\"],data_exons_minus[\"transcriptID\"]]), \"gene\":pd.concat([data_exons_plus[\"gene\"],data_exons_minus[\"gene\"]])},\n",
    "                       columns=[\"chrm\",\"start\",\"end\",\"transcriptID\",\"strand\",\"gene\"])\n",
    "print data_5p.head()\n",
    "print data_5p.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   chrm  start    end    transcriptID strand         gene\n",
      "0  chr1  11774  11974     NR_046018.2      +      DDX11L1\n",
      "1  chr1  12513  12713     NR_046018.2      +      DDX11L1\n",
      "2  chr1  13121  13321     NR_046018.2      +      DDX11L1\n",
      "3  chr1  29826  30026  XR_001737835.1      +  MIR1302-2HG\n",
      "4  chr1  30266  30466     NR_036051.1      +    MIR1302-2\n",
      "        chrm      start        end    transcriptID strand          gene\n",
      "919719  chr9  138044971  138045171     NR_121582.1      -  LOC101928786\n",
      "919720  chr9  138200182  138200382     NR_135133.1      -  LOC101928932\n",
      "919721  chr9  138203353  138203553     NR_135133.1      -  LOC101928932\n",
      "919722  chr9  138216329  138216529  XR_001746987.2      -  LOC107987145\n",
      "919723  chr9  138216553  138216753  XR_001746987.2      -  LOC107987145\n"
     ]
    }
   ],
   "source": [
    "# Get 3' exon-intron boundaries\n",
    "data_3p = pd.DataFrame({\"chrm\":pd.concat([data_exons_plus[\"chrm\"],data_exons_minus[\"chrm\"]]),\"start\":pd.concat([data_exons_plus[\"start\"]-100,data_exons_minus[\"end\"]-100]),\n",
    "                       \"end\":pd.concat([data_exons_plus[\"start\"]+100,data_exons_minus[\"end\"]+100]), \"strand\":pd.concat([data_exons_plus[\"strand\"],data_exons_minus[\"strand\"]]),\n",
    "                       \"transcriptID\":pd.concat([data_exons_plus[\"transcriptID\"],data_exons_minus[\"transcriptID\"]]), \"gene\":pd.concat([data_exons_plus[\"gene\"],data_exons_minus[\"gene\"]])},\n",
    "                       columns=[\"chrm\",\"start\",\"end\",\"transcriptID\",\"strand\",\"gene\"])\n",
    "print data_3p.head()\n",
    "print data_3p.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-37-b98117561945>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     29\u001b[0m     \u001b[0mall_5p_ExonIntronCoords_ToWrite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 31\u001b[0;31m     \u001b[0mexonintron_coords_3p\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdata_3p_groupedByTranscript\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_group\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     32\u001b[0m     \u001b[0mexonintron_coords_3p_sorted\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexonintron_coords_3p\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"chrm\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"start\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     33\u001b[0m     \u001b[0;31m# We are getting rid of the last row for 3p, because that is not actually including the intron, it's the genomic region\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# 5p exon-intron boundaries grouped\n",
    "data_5p_groupedByTranscript = data_5p.groupby(by=[\"transcriptID\"])\n",
    "# 3p exon-intron boundaries grouped\n",
    "data_3p_groupedByTranscript = data_3p.groupby(by=[\"transcriptID\"])\n",
    "# All exons grouped\n",
    "data_exons_groupedByTranscript = data_exons.groupby(by=[\"transcriptID\"])\n",
    "\n",
    "all_5p_ExonIntronCoords_ToWrite = pd.DataFrame(columns=[\"chrm\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "all_3p_ExonIntronCoords_ToWrite = pd.DataFrame(columns=[\"chrm\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "new_geneNameDF = pd.DataFrame(columns=[\"TranscriptID\",\"Gene\"])\n",
    "\n",
    "for key, item in data_exons_groupedByTranscript:\n",
    "    \n",
    "    coords_key = data_exons_groupedByTranscript.get_group(key)\n",
    "    num_rows = coords_key.shape[0]\n",
    "    transcriptIDs = [key+\"_\"+str(i+1) for i in range(num_rows-1)]\n",
    "    \n",
    "    exonintron_coords_5p = data_5p_groupedByTranscript.get_group(key)\n",
    "    exonintron_coords_5p_sorted = exonintron_coords_5p.sort_values(by=[\"chrm\",\"start\"])\n",
    "    # We are getting rid of the first row for 5p, because that is not actually including the intron, it's the genomic region\n",
    "    #print list(exonintron_coords_5p_sorted[\"strand\"].values)\n",
    "    if list(exonintron_coords_5p_sorted[\"strand\"].values)[0]==\"+\":\n",
    "        exonintron_coords_5p_sorted = exonintron_coords_5p_sorted.iloc[:-1]\n",
    "    else:\n",
    "        exonintron_coords_5p_sorted = exonintron_coords_5p_sorted.iloc[1:]\n",
    "    new_DF_5p = pd.DataFrame({\"chrm\":exonintron_coords_5p_sorted[\"chrm\"],\"start\":exonintron_coords_5p_sorted[\"start\"],\"end\":exonintron_coords_5p_sorted[\"end\"],\"transcriptID\":transcriptIDs,\"strand\":exonintron_coords_5p_sorted[\"strand\"]}\n",
    "                          ,columns=[\"chrm\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "    all_5p_ExonIntronCoords_ToWrite = pd.concat([all_5p_ExonIntronCoords_ToWrite,new_DF_5p],ignore_index=True,axis=0)\n",
    "    all_5p_ExonIntronCoords_ToWrite.reset_index(drop=True)\n",
    "    \n",
    "    exonintron_coords_3p = data_3p_groupedByTranscript.get_group(key)\n",
    "    exonintron_coords_3p_sorted = exonintron_coords_3p.sort_values(by=[\"chrm\",\"start\"])\n",
    "    # We are getting rid of the last row for 3p, because that is not actually including the intron, it's the genomic region\n",
    "    if list(exonintron_coords_3p_sorted[\"strand\"].values)[0]==\"+\":\n",
    "        exonintron_coords_3p_sorted = exonintron_coords_3p_sorted.iloc[1:]\n",
    "    else:\n",
    "        exonintron_coords_3p_sorted = exonintron_coords_3p_sorted.iloc[:-1]\n",
    "    new_DF_3p = pd.DataFrame({\"chrm\":exonintron_coords_3p_sorted[\"chrm\"],\"start\":exonintron_coords_3p_sorted[\"start\"],\"end\":exonintron_coords_3p_sorted[\"end\"],\"transcriptID\":transcriptIDs,\"strand\":exonintron_coords_3p_sorted[\"strand\"]}\n",
    "                          ,columns=[\"chrm\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "    all_3p_ExonIntronCoords_ToWrite = pd.concat([all_3p_ExonIntronCoords_ToWrite,new_DF_3p],ignore_index=True,axis=0)\n",
    "    all_3p_ExonIntronCoords_ToWrite.reset_index(drop=True)\n",
    "    \n",
    "    new_geneNameDF = pd.concat([new_geneNameDF,pd.DataFrame({\"Gene\":list(coords_key[\"gene\"].values)[1:],\"TranscriptID\":transcriptIDs},columns=[\"TranscriptID\",\"Gene\"])],ignore_index=True,axis=0)\n",
    "    new_geneNameDF.reset_index(drop=True)\n",
    "    \n",
    "new_geneNameDF.to_csv(\"../data/ExonIntronCoords_GeneNameTranscriptID_NCBI_RefSeq_hg38_FromGFFfile.tsv\",sep=\"\\t\",header=False,index=False)\n",
    "\n",
    "all_5p_ExonIntronCoords_ToWrite[\"chrm\"] = all_5p_ExonIntronCoords_ToWrite[\"chrm\"].replace(\"chr23\",\"chrX\")\n",
    "all_5p_ExonIntronCoords_ToWrite[\"chrm\"] = all_5p_ExonIntronCoords_ToWrite[\"chrm\"].replace(\"chr24\",\"chrY\")\n",
    "\n",
    "\n",
    "all_5p_ExonIntronCoords_ToWrite_Sorted = all_5p_ExonIntronCoords_ToWrite.sort_values(by=[\"chrm\",\"start\"])\n",
    "all_5p_ExonIntronCoords_ToWrite_Sorted.to_csv(\"../data/ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence.bed\",sep=\"\\t\",header=False,index=False)\n",
    "\n",
    "all_3p_ExonIntronCoords_ToWrite[\"chrm\"] = all_3p_ExonIntronCoords_ToWrite[\"chrm\"].replace(\"chr23\",\"chrX\")\n",
    "all_3p_ExonIntronCoords_ToWrite[\"chrm\"] = all_3p_ExonIntronCoords_ToWrite[\"chrm\"].replace(\"chr24\",\"chrY\")\n",
    "\n",
    "\n",
    "all_3p_ExonIntronCoords_ToWrite_Sorted = all_3p_ExonIntronCoords_ToWrite.sort_values(by=[\"chrm\",\"start\"])\n",
    "all_3p_ExonIntronCoords_ToWrite_Sorted.to_csv(\"../data/ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63426</th>\n",
       "      <td>chr1</td>\n",
       "      <td>999426</td>\n",
       "      <td>999626</td>\n",
       "      <td>NM_001142467.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63427</th>\n",
       "      <td>chr1</td>\n",
       "      <td>999592</td>\n",
       "      <td>999792</td>\n",
       "      <td>NM_001142467.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53638</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1173826</td>\n",
       "      <td>1174026</td>\n",
       "      <td>NM_001130045.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53639</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1174221</td>\n",
       "      <td>1174421</td>\n",
       "      <td>NM_001130045.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53640</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1174389</td>\n",
       "      <td>1174589</td>\n",
       "      <td>NM_001130045.1_3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrm    start      end      transcriptID strand\n",
       "63426  chr1   999426   999626  NM_001142467.1_1      -\n",
       "63427  chr1   999592   999792  NM_001142467.1_2      -\n",
       "53638  chr1  1173826  1174026  NM_001130045.1_1      +\n",
       "53639  chr1  1174221  1174421  NM_001130045.1_2      +\n",
       "53640  chr1  1174389  1174589  NM_001130045.1_3      +"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_5p_ExonIntronCoords_ToWrite.sort_values(by=[\"chrm\",\"start\"]).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>63426</th>\n",
       "      <td>chr1</td>\n",
       "      <td>999332</td>\n",
       "      <td>999532</td>\n",
       "      <td>NM_001142467.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63427</th>\n",
       "      <td>chr1</td>\n",
       "      <td>999513</td>\n",
       "      <td>999713</td>\n",
       "      <td>NM_001142467.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53638</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1174185</td>\n",
       "      <td>1174385</td>\n",
       "      <td>NM_001130045.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53639</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1174324</td>\n",
       "      <td>1174524</td>\n",
       "      <td>NM_001130045.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53640</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1179089</td>\n",
       "      <td>1179289</td>\n",
       "      <td>NM_001130045.1_3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53641</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1179557</td>\n",
       "      <td>1179757</td>\n",
       "      <td>NM_001130045.1_4</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53642</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1179934</td>\n",
       "      <td>1180134</td>\n",
       "      <td>NM_001130045.1_5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53643</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1180383</td>\n",
       "      <td>1180583</td>\n",
       "      <td>NM_001130045.1_6</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53644</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1180631</td>\n",
       "      <td>1180831</td>\n",
       "      <td>NM_001130045.1_7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53645</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1181641</td>\n",
       "      <td>1181841</td>\n",
       "      <td>NM_001130045.1_8</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53646</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1182261</td>\n",
       "      <td>1182461</td>\n",
       "      <td>NM_001130045.1_9</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53647</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1182776</td>\n",
       "      <td>1182976</td>\n",
       "      <td>NM_001130045.1_10</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53648</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1183820</td>\n",
       "      <td>1184020</td>\n",
       "      <td>NM_001130045.1_11</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53649</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1184869</td>\n",
       "      <td>1185069</td>\n",
       "      <td>NM_001130045.1_12</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53650</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1196500</td>\n",
       "      <td>1196700</td>\n",
       "      <td>NM_001130045.1_13</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53651</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1196993</td>\n",
       "      <td>1197193</td>\n",
       "      <td>NM_001130045.1_14</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53652</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1197338</td>\n",
       "      <td>1197538</td>\n",
       "      <td>NM_001130045.1_15</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20232</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1242546</td>\n",
       "      <td>1242746</td>\n",
       "      <td>NM_001014980.2_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20233</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1242813</td>\n",
       "      <td>1243013</td>\n",
       "      <td>NM_001014980.2_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20234</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1243052</td>\n",
       "      <td>1243252</td>\n",
       "      <td>NM_001014980.2_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20235</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1243452</td>\n",
       "      <td>1243652</td>\n",
       "      <td>NM_001014980.2_4</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20236</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1244005</td>\n",
       "      <td>1244205</td>\n",
       "      <td>NM_001014980.2_5</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20237</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1244175</td>\n",
       "      <td>1244375</td>\n",
       "      <td>NM_001014980.2_6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20238</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1244397</td>\n",
       "      <td>1244597</td>\n",
       "      <td>NM_001014980.2_7</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54590</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1281126</td>\n",
       "      <td>1281326</td>\n",
       "      <td>NM_001130413.3_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54591</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1281311</td>\n",
       "      <td>1281511</td>\n",
       "      <td>NM_001130413.3_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54592</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1282142</td>\n",
       "      <td>1282342</td>\n",
       "      <td>NM_001130413.3_3</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54593</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1283878</td>\n",
       "      <td>1284078</td>\n",
       "      <td>NM_001130413.3_4</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54594</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1285471</td>\n",
       "      <td>1285671</td>\n",
       "      <td>NM_001130413.3_5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54595</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1285826</td>\n",
       "      <td>1286026</td>\n",
       "      <td>NM_001130413.3_6</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54596</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1286668</td>\n",
       "      <td>1286868</td>\n",
       "      <td>NM_001130413.3_7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54597</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1287009</td>\n",
       "      <td>1287209</td>\n",
       "      <td>NM_001130413.3_8</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54598</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1287408</td>\n",
       "      <td>1287608</td>\n",
       "      <td>NM_001130413.3_9</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54599</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1287573</td>\n",
       "      <td>1287773</td>\n",
       "      <td>NM_001130413.3_10</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54600</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1287839</td>\n",
       "      <td>1288039</td>\n",
       "      <td>NM_001130413.3_11</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54601</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1290171</td>\n",
       "      <td>1290371</td>\n",
       "      <td>NM_001130413.3_12</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54602</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1290377</td>\n",
       "      <td>1290577</td>\n",
       "      <td>NM_001130413.3_13</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54603</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1290537</td>\n",
       "      <td>1290737</td>\n",
       "      <td>NM_001130413.3_14</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54604</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1290795</td>\n",
       "      <td>1290995</td>\n",
       "      <td>NM_001130413.3_15</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54605</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1290965</td>\n",
       "      <td>1291165</td>\n",
       "      <td>NM_001130413.3_16</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54606</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1291154</td>\n",
       "      <td>1291354</td>\n",
       "      <td>NM_001130413.3_17</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23721</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1326736</td>\n",
       "      <td>1326936</td>\n",
       "      <td>NM_001029885.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23722</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1327141</td>\n",
       "      <td>1327341</td>\n",
       "      <td>NM_001029885.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48998</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1373802</td>\n",
       "      <td>1374002</td>\n",
       "      <td>NM_001127229.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49001</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1373802</td>\n",
       "      <td>1374002</td>\n",
       "      <td>NM_001127230.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48999</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1374345</td>\n",
       "      <td>1374545</td>\n",
       "      <td>NM_001127229.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49002</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1374345</td>\n",
       "      <td>1374545</td>\n",
       "      <td>NM_001127230.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49000</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1374690</td>\n",
       "      <td>1374890</td>\n",
       "      <td>NM_001127229.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49003</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1374690</td>\n",
       "      <td>1374890</td>\n",
       "      <td>NM_001127230.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27749</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1392703</td>\n",
       "      <td>1392903</td>\n",
       "      <td>NM_001039577.4_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrm    start      end       transcriptID strand\n",
       "63426  chr1   999332   999532   NM_001142467.1_1      -\n",
       "63427  chr1   999513   999713   NM_001142467.1_2      -\n",
       "53638  chr1  1174185  1174385   NM_001130045.1_1      +\n",
       "53639  chr1  1174324  1174524   NM_001130045.1_2      +\n",
       "53640  chr1  1179089  1179289   NM_001130045.1_3      +\n",
       "53641  chr1  1179557  1179757   NM_001130045.1_4      +\n",
       "53642  chr1  1179934  1180134   NM_001130045.1_5      +\n",
       "53643  chr1  1180383  1180583   NM_001130045.1_6      +\n",
       "53644  chr1  1180631  1180831   NM_001130045.1_7      +\n",
       "53645  chr1  1181641  1181841   NM_001130045.1_8      +\n",
       "53646  chr1  1182261  1182461   NM_001130045.1_9      +\n",
       "53647  chr1  1182776  1182976  NM_001130045.1_10      +\n",
       "53648  chr1  1183820  1184020  NM_001130045.1_11      +\n",
       "53649  chr1  1184869  1185069  NM_001130045.1_12      +\n",
       "53650  chr1  1196500  1196700  NM_001130045.1_13      +\n",
       "53651  chr1  1196993  1197193  NM_001130045.1_14      +\n",
       "53652  chr1  1197338  1197538  NM_001130045.1_15      +\n",
       "20232  chr1  1242546  1242746   NM_001014980.2_1      -\n",
       "20233  chr1  1242813  1243013   NM_001014980.2_2      -\n",
       "20234  chr1  1243052  1243252   NM_001014980.2_3      -\n",
       "20235  chr1  1243452  1243652   NM_001014980.2_4      -\n",
       "20236  chr1  1244005  1244205   NM_001014980.2_5      -\n",
       "20237  chr1  1244175  1244375   NM_001014980.2_6      -\n",
       "20238  chr1  1244397  1244597   NM_001014980.2_7      -\n",
       "54590  chr1  1281126  1281326   NM_001130413.3_1      +\n",
       "54591  chr1  1281311  1281511   NM_001130413.3_2      +\n",
       "54592  chr1  1282142  1282342   NM_001130413.3_3      +\n",
       "54593  chr1  1283878  1284078   NM_001130413.3_4      +\n",
       "54594  chr1  1285471  1285671   NM_001130413.3_5      +\n",
       "54595  chr1  1285826  1286026   NM_001130413.3_6      +\n",
       "54596  chr1  1286668  1286868   NM_001130413.3_7      +\n",
       "54597  chr1  1287009  1287209   NM_001130413.3_8      +\n",
       "54598  chr1  1287408  1287608   NM_001130413.3_9      +\n",
       "54599  chr1  1287573  1287773  NM_001130413.3_10      +\n",
       "54600  chr1  1287839  1288039  NM_001130413.3_11      +\n",
       "54601  chr1  1290171  1290371  NM_001130413.3_12      +\n",
       "54602  chr1  1290377  1290577  NM_001130413.3_13      +\n",
       "54603  chr1  1290537  1290737  NM_001130413.3_14      +\n",
       "54604  chr1  1290795  1290995  NM_001130413.3_15      +\n",
       "54605  chr1  1290965  1291165  NM_001130413.3_16      +\n",
       "54606  chr1  1291154  1291354  NM_001130413.3_17      +\n",
       "23721  chr1  1326736  1326936   NM_001029885.1_1      +\n",
       "23722  chr1  1327141  1327341   NM_001029885.1_2      +\n",
       "48998  chr1  1373802  1374002   NM_001127229.1_1      -\n",
       "49001  chr1  1373802  1374002   NM_001127230.1_1      -\n",
       "48999  chr1  1374345  1374545   NM_001127229.1_2      -\n",
       "49002  chr1  1374345  1374545   NM_001127230.1_2      -\n",
       "49000  chr1  1374690  1374890   NM_001127229.1_3      -\n",
       "49003  chr1  1374690  1374890   NM_001127230.1_3      -\n",
       "27749  chr1  1392703  1392903   NM_001039577.4_1      -"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_3p_ExonIntronCoords_ToWrite.sort_values(by=[\"chrm\",\"start\"]).head(n=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12513  12713  NR_046018.2_1  +\n",
       "1  chr1  13121  13321  NR_046018.2_2  +\n",
       "2  chr1  14729  14929  NR_024540.1_1  -\n",
       "3  chr1  14938  15138  NR_024540.1_2  -\n",
       "4  chr1  15847  16047  NR_024540.1_3  -"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_3p_ExonIntronCoords_ToWrite_Sorted = pd.read_csv(\"../data/ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence_Just3p.bed\",sep=\"\\t\",header=None)\n",
    "all_3p_ExonIntronCoords_ToWrite_Sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12127  12327  NR_046018.2_1  +\n",
       "1  chr1  12621  12821  NR_046018.2_2  +\n",
       "2  chr1  14870  15070  NR_024540.1_1  -\n",
       "3  chr1  15696  15896  NR_024540.1_2  -\n",
       "4  chr1  16507  16707  NR_024540.1_3  -"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_5p_ExonIntronCoords_ToWrite_Sorted = pd.read_csv(\"../data/ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence_Just5p.bed\",sep=\"\\t\",header=None)\n",
    "all_5p_ExonIntronCoords_ToWrite_Sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12127  12327  NR_046018.2_1  +\n",
       "1  chr1  12621  12821  NR_046018.2_2  +"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_5p_ExonIntronCoords_ToWrite_Sorted[all_5p_ExonIntronCoords_ToWrite_Sorted[3].str.contains(\"NR_046018.2\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "bedtools getfasta -name -s -fo ../data/ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence_Just5p.fa -fi /home/shared/GenomeReference/hg38.fa -bed ../data/ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence_Just5p.bed\n",
    "bedtools getfasta -name -s -fo ../data/ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence_Just3p.fa -fi /home/shared/GenomeReference/hg38.fa -bed ../data/ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence_Just3p.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n",
      "(273789, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12127  12327  NR_046018.2_1  +\n",
       "1  chr1  12621  12821  NR_046018.2_2  +\n",
       "2  chr1  14870  15070  NR_024540.1_1  -\n",
       "3  chr1  15696  15896  NR_024540.1_2  -\n",
       "4  chr1  16507  16707  NR_024540.1_3  -"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's get the unique coordinates \n",
    "all_5p_ExonIntronCoords_ToWrite_Sorted_unique = all_5p_ExonIntronCoords_ToWrite_Sorted.drop_duplicates(subset=[0,1,2])\n",
    "print all_5p_ExonIntronCoords_ToWrite_Sorted.shape\n",
    "print all_5p_ExonIntronCoords_ToWrite_Sorted_unique.shape\n",
    "all_5p_ExonIntronCoords_ToWrite_Sorted_unique.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n",
      "(265896, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12513  12713  NR_046018.2_1  +\n",
       "1  chr1  13121  13321  NR_046018.2_2  +\n",
       "2  chr1  14729  14929  NR_024540.1_1  -\n",
       "3  chr1  14938  15138  NR_024540.1_2  -\n",
       "4  chr1  15847  16047  NR_024540.1_3  -"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's get the unique coordinates \n",
    "all_3p_ExonIntronCoords_ToWrite_Sorted_unique = all_3p_ExonIntronCoords_ToWrite_Sorted.drop_duplicates(subset=[0,1,2])\n",
    "print all_3p_ExonIntronCoords_ToWrite_Sorted.shape\n",
    "print all_3p_ExonIntronCoords_ToWrite_Sorted_unique.shape\n",
    "all_3p_ExonIntronCoords_ToWrite_Sorted_unique.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write these coordinates to a file\n",
    "all_5p_ExonIntronCoords_ToWrite_Sorted_unique.to_csv(\"../data/Unique_ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence_Just5p.bed\",sep=\"\\t\",header=False,index=False)\n",
    "all_3p_ExonIntronCoords_ToWrite_Sorted_unique.to_csv(\"../data/Unique_ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence_Just3p.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "bedtools getfasta -name -s -fo ../data/Unique_ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence_Just5p.fa -fi /home/shared/GenomeReference/hg38.fa -bed ../data/Unique_ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence_Just5p.bed\n",
    "bedtools getfasta -name -s -fo ../data/Unique_ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence_Just3p.fa -fi /home/shared/GenomeReference/hg38.fa -bed ../data/Unique_ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence_Just3p.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Let's get the intron coordinates from this exon data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's group by the transcript ID\n",
    "data_exons_GrpedByFeature = data_exons.groupby(by=[\"transcriptID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allintrons_DF = pd.DataFrame(columns=[\"chrm\",\"start\",\"end\",\"strand\",\"gene\",\"transcriptID\"])\n",
    "for key, item in data_exons_GrpedByFeature:\n",
    "    exons = data_exons_GrpedByFeature.get_group(key)\n",
    "    # Only do this if there are more than 1 exon\n",
    "    if exons.shape[0] > 1:\n",
    "        clump_start_end = sorted(list(exons[\"start\"].values) + list(exons[\"end\"].values))\n",
    "        # Get start coordinates of introns \n",
    "        start_coords_intron = [clump_start_end[i] for i in range(1,len(clump_start_end)-2,2)]\n",
    "        # Get end coordinates of introns\n",
    "        end_coords_intron = [clump_start_end[i]-1 for i in range(2,len(clump_start_end)-1,2)]\n",
    "        if len(start_coords_intron)==len(end_coords_intron):\n",
    "            chrm = [list(exons[\"chrm\"].values)[0]]*len(start_coords_intron)\n",
    "            gene = [list(exons[\"gene\"].values)[0]]*len(start_coords_intron)\n",
    "            strand = [list(exons[\"strand\"].values)[0]]*len(start_coords_intron)\n",
    "            transcriptID = [key]*len(start_coords_intron)\n",
    "            newintron_DF = pd.DataFrame({\"chrm\":chrm,\"start\":start_coords_intron,\"end\":end_coords_intron,\"strand\":strand,\"transcriptID\":transcriptID,\"gene\":gene},columns=[\"chrm\",\"start\",\"end\",\"strand\",\"gene\",\"transcriptID\"])\n",
    "            allintrons_DF = pd.concat([allintrons_DF,newintron_DF],ignore_index=True,axis=0)\n",
    "            allintrons_DF = allintrons_DF.reset_index(drop=True)\n",
    "        else:\n",
    "            print key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allintrons_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allintrons_DF.to_csv(\"../data/ExtractedIntronCoordinatesFromGFFfile.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allintrons_DF = pd.read_csv(\"../data/ExtractedIntronCoordinatesFromGFFfile.bed\",sep=\"\\t\",header=None)\n",
    "print allintrons_DF.shape\n",
    "allintrons_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allintrons_DF_groupedByTranscript = allintrons_DF.groupby(by=[5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allintrons_DF_groupedByTranscript.get_group(\"NM_000014.5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allIntronicCoords_ToWrite = pd.DataFrame(columns=[\"chrm\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "new_geneNameDF = pd.DataFrame(columns=[\"TranscriptID\",\"Gene\"])\n",
    "for key, item in allintrons_DF_groupedByTranscript:\n",
    "    introns = allintrons_DF_groupedByTranscript.get_group(key)\n",
    "    introns_sorted = introns.sort_values(by=[0,1])\n",
    "    num_rows = introns.shape[0]\n",
    "    transcriptIDs = [key+\"_\"+str(i+1) for i in range(num_rows)]\n",
    "    new_DF = pd.DataFrame({\"chrm\":introns_sorted[0],\"start\":introns_sorted[1],\"end\":introns_sorted[2],\"transcriptID\":transcriptIDs,\"strand\":introns_sorted[3]},columns=[\"chrm\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "    allIntronicCoords_ToWrite = pd.concat([allIntronicCoords_ToWrite,new_DF],ignore_index=True,axis=0)\n",
    "    allIntronicCoords_ToWrite.reset_index(drop=True)\n",
    "    new_geneNameDF = pd.concat([new_geneNameDF,pd.DataFrame({\"Gene\":introns_sorted[4],\"TranscriptID\":transcriptIDs},columns=[\"TranscriptID\",\"Gene\"])],ignore_index=True,axis=0)\n",
    "    new_geneNameDF.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allIntronicCoords_ToWrite = pd.read_csv(\"../data/ExtractedIntronCoordinatesFromGFFfile_ForSequence.bed\",sep=\"\\t\",header=None)\n",
    "allIntronicCoords_ToWrite.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_geneNameDF = pd.read_csv(\"../data/JustIntrons_GeneNameTranscriptID_NCBI_RefSeq_hg38_FromGFFfile.tsv\",sep=\"\\t\",header=None)\n",
    "new_geneNameDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "bedtools getfasta -name -s -fo ../data/ExtractedIntronCoordinatesFromGFFfile_sequences.fa -fi /home/shared/GenomeReference/hg38.fa -bed ../data/ExtractedIntronCoordinatesFromGFFfile_ForSequence.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Let's get the genomic coordinates of transcripts from this exon data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's extract the entire genomic coordinate\n",
    "allGenomicCoords_DF = pd.DataFrame(columns=[\"chrm\",\"start\",\"end\",\"strand\",\"gene\",\"transcriptID\"])\n",
    "for key, item in data_exons_GrpedByFeature:\n",
    "    exons = data_exons_GrpedByFeature.get_group(key)\n",
    "    chrm = [list(exons[\"chrm\"].values)[0]]\n",
    "    gene = [list(exons[\"gene\"].values)[0]]\n",
    "    strand = [list(exons[\"strand\"].values)[0]]\n",
    "    transcriptID = [key]\n",
    "    clump_start_end = sorted(list(exons[\"start\"].values) + list(exons[\"end\"].values))\n",
    "    start = [clump_start_end[0]]\n",
    "    end = [clump_start_end[len(clump_start_end)-1]]\n",
    "    # Only do this if there are more than 1 exon\n",
    "    newGenomicCoord_DF = pd.DataFrame({\"chrm\":chrm,\"start\":start,\"end\":end,\"strand\":strand,\"transcriptID\":transcriptID,\"gene\":gene},columns=[\"chrm\",\"start\",\"end\",\"strand\",\"gene\",\"transcriptID\"])\n",
    "    allGenomicCoords_DF = pd.concat([allGenomicCoords_DF,newGenomicCoord_DF],ignore_index=True,axis=0)\n",
    "    allGenomicCoords_DF = allGenomicCoords_DF.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allGenomicCoords_DF.to_csv(\"../data/ExtractedGenomicCoordinatesFromGFFfile.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allGenomicCoords_DF = pd.read_csv(\"../data/ExtractedGenomicCoordinatesFromGFFfile.bed\",sep=\"\\t\",header=None)\n",
    "print allGenomicCoords_DF.shape\n",
    "allGenomicCoords_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allGenomicCoords_DF_sorted = allGenomicCoords_DF.sort_values(by=[0,1])\n",
    "allGenomicCoords_DF_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allGenomicCoords_DF_sorted[0] = allGenomicCoords_DF_sorted[0].replace(\"chr23\",\"chrX\")\n",
    "allGenomicCoords_DF_sorted[0] = allGenomicCoords_DF_sorted[0].replace(\"chr24\",\"chrY\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allGenomicCoords_DF_sorted[[0,1,2,5,3]].to_csv(\"../data/ExtractedGenomicCoordinatesFromGFFfile_Sorted.bed\",sep=\"\\t\",header=False,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%script bash\n",
    "bedtools getfasta -name -s -fo ../data/ExtractedGenomicCoordinatesFromGFFfile_sequences.fa -fi /home/shared/GenomeReference/hg38.fa -bed ../data/ExtractedGenomicCoordinatesFromGFFfile_Sorted.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Let's get the coordinates and gene ID names to make simplified annotation format (SAF) for featureCounts\n",
    "\n",
    "### First for transcriptome coordinates, we need both geneID and transcriptID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2017619, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2  3        4            5           6           7\n",
       "0  chr1  11874  12227  +  DDX11L1  NR_046018.2  BestRefSeq        exon\n",
       "1  chr1  11874  14409  +  DDX11L1  NR_046018.2  BestRefSeq  transcript\n",
       "2  chr1  12613  12721  +  DDX11L1  NR_046018.2  BestRefSeq        exon\n",
       "3  chr1  13221  14409  +  DDX11L1  NR_046018.2  BestRefSeq        exon\n",
       "4  chr1  14362  14829  -   WASH7P  NR_024540.1  BestRefSeq        exon"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_genes = pd.read_csv(\"../data/GRCh38_latest_genomic_ValidChroms_OnlyGenes.bed\",sep=\"\\t\",header=None)\n",
    "print data_genes.shape\n",
    "data_genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set column names\n",
    "data_genes.columns = [\"chrm\",\"start\",\"end\",\"strand\",\"gene\",\"transcriptID\",\"RefSeqType\",\"feature\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1859728, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrm</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>RefSeqType</th>\n",
       "      <th>feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>-</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrm  start    end strand     gene transcriptID  RefSeqType feature\n",
       "0  chr1  11874  12227      +  DDX11L1  NR_046018.2  BestRefSeq    exon\n",
       "2  chr1  12613  12721      +  DDX11L1  NR_046018.2  BestRefSeq    exon\n",
       "3  chr1  13221  14409      +  DDX11L1  NR_046018.2  BestRefSeq    exon\n",
       "4  chr1  14362  14829      -   WASH7P  NR_024540.1  BestRefSeq    exon\n",
       "6  chr1  14970  15038      -   WASH7P  NR_024540.1  BestRefSeq    exon"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only grab exons \n",
    "data_exons = data_genes[data_genes[\"feature\"]==\"exon\"]\n",
    "print data_exons.shape\n",
    "data_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2  chr1  11874  12227      +\n",
       "2  NR_046018.2  chr1  12613  12721      +\n",
       "3  NR_046018.2  chr1  13221  14409      +\n",
       "4  NR_024540.1  chr1  14362  14829      -\n",
       "6  NR_024540.1  chr1  14970  15038      -"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's write a file that is SAF for just transcriptIDs\n",
    "data_exons_justTranscriptIDs = data_exons[[\"transcriptID\",\"chrm\",\"start\",\"end\",\"strand\"]]\n",
    "data_exons_justTranscriptIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_exons_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_exons_justTranscriptIDs.to_csv(\"../data/SAFformat_GRCh38_latest_genomic_ValidChroms_OnlyTranscripts.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>14829</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14970</td>\n",
       "      <td>15038</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GeneID   Chr  Start    End Strand\n",
       "0  DDX11L1  chr1  11874  12227      +\n",
       "2  DDX11L1  chr1  12613  12721      +\n",
       "3  DDX11L1  chr1  13221  14409      +\n",
       "4   WASH7P  chr1  14362  14829      -\n",
       "6   WASH7P  chr1  14970  15038      -"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's write a file that is SAF for just genes\n",
    "data_exons_justGeneIDs = data_exons[[\"gene\",\"chrm\",\"start\",\"end\",\"strand\"]]\n",
    "data_exons_justGeneIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_exons_justGeneIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_exons_justGeneIDs.to_csv(\"../data/SAFformat_GRCh38_latest_genomic_ValidChroms_OnlyGenes.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Next for genomic coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(159998, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_106918.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XR_001737835.1</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_036051.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                0            1\n",
       "0     NR_046018.2      DDX11L1\n",
       "1     NR_024540.1       WASH7P\n",
       "2     NR_106918.1    MIR6859-1\n",
       "3  XR_001737835.1  MIR1302-2HG\n",
       "4     NR_036051.1    MIR1302-2"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### First for genome coordinates, we need both geneID and transcriptID\n",
    "data_genes_transcripts = pd.read_csv(\"../data/GeneNameTranscriptID_NCBI_RefSeq_hg38_FromGFFfile.tsv\",sep=\"\\t\",header=None)\n",
    "print data_genes_transcripts.shape\n",
    "data_genes_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157829, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NR_106918.1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29926</td>\n",
       "      <td>31295</td>\n",
       "      <td>XR_001737835.1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>NR_036051.1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2               3  4\n",
       "0  chr1  11874  14409     NR_046018.2  +\n",
       "1  chr1  14362  29370     NR_024540.1  -\n",
       "2  chr1  17369  17436     NR_106918.1  -\n",
       "3  chr1  29926  31295  XR_001737835.1  +\n",
       "4  chr1  30366  30503     NR_036051.1  +"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coords_genomic = pd.read_csv(\"../data/ExtractedGenomicCoordinatesFromGFFfile_Sorted.bed\",sep=\"\\t\",header=None)\n",
    "print coords_genomic.shape\n",
    "coords_genomic.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157829, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_046018.2</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>-</td>\n",
       "      <td>NR_024540.1</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>NR_106918.1</td>\n",
       "      <td>-</td>\n",
       "      <td>NR_106918.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29926</td>\n",
       "      <td>31295</td>\n",
       "      <td>XR_001737835.1</td>\n",
       "      <td>+</td>\n",
       "      <td>XR_001737835.1</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>NR_036051.1</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_036051.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0_x    1_x      2               3  4             0_y          1_y\n",
       "0  chr1  11874  14409     NR_046018.2  +     NR_046018.2      DDX11L1\n",
       "1  chr1  14362  29370     NR_024540.1  -     NR_024540.1       WASH7P\n",
       "2  chr1  17369  17436     NR_106918.1  -     NR_106918.1    MIR6859-1\n",
       "3  chr1  29926  31295  XR_001737835.1  +  XR_001737835.1  MIR1302-2HG\n",
       "4  chr1  30366  30503     NR_036051.1  +     NR_036051.1    MIR1302-2"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_coords_genomic = coords_genomic.merge(data_genes_transcripts,right_on=0,left_on=3)\n",
    "print merged_coords_genomic.shape\n",
    "merged_coords_genomic.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(157829, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1_y</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29926</td>\n",
       "      <td>31295</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           1_y   0_x    1_x      2  4\n",
       "0      DDX11L1  chr1  11874  14409  +\n",
       "1       WASH7P  chr1  14362  29370  -\n",
       "2    MIR6859-1  chr1  17369  17436  -\n",
       "3  MIR1302-2HG  chr1  29926  31295  +\n",
       "4    MIR1302-2  chr1  30366  30503  +"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the SAF format\n",
    "data_genes_justGeneIDs = merged_coords_genomic.iloc[:,[6,0,1,2,4]]\n",
    "print data_genes_justGeneIDs.shape\n",
    "data_genes_justGeneIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29926</td>\n",
       "      <td>31295</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GeneID   Chr  Start    End Strand\n",
       "0      DDX11L1  chr1  11874  14409      +\n",
       "1       WASH7P  chr1  14362  29370      -\n",
       "2    MIR6859-1  chr1  17369  17436      -\n",
       "3  MIR1302-2HG  chr1  29926  31295      +\n",
       "4    MIR1302-2  chr1  30366  30503      +"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_genes_justGeneIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_genes_justGeneIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29926</td>\n",
       "      <td>31295</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        GeneID   Chr  Start    End Strand\n",
       "0      DDX11L1  chr1  11874  14409      +\n",
       "1       WASH7P  chr1  14362  29370      -\n",
       "2    MIR6859-1  chr1  17369  17436      -\n",
       "3  MIR1302-2HG  chr1  29926  31295      +\n",
       "4    MIR1302-2  chr1  30366  30503      +"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_genes_justGeneIDs_sorted = data_genes_justGeneIDs.sort_values(by=[\"Chr\",\"Start\",\"End\"])\n",
    "data_genes_justGeneIDs_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_genes_justGeneIDs_sorted.to_csv(\"../data/SAFformat_ExtractedGenomicCoordinatesFromGFFfile_OnlyGenes.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Next for exon-intron coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12127  12327  NR_046018.2_1  +\n",
       "1  chr1  12621  12821  NR_046018.2_2  +\n",
       "2  chr1  14870  15070  NR_024540.1_1  -\n",
       "3  chr1  15696  15896  NR_024540.1_2  -\n",
       "4  chr1  16507  16707  NR_024540.1_3  -"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### First for exon-intron coordinates, we need both geneID and transcriptID\n",
    "data_transcripts = pd.read_csv(\"../data/ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence_Just5p.bed\",sep=\"\\t\",header=None)\n",
    "print data_transcripts.shape\n",
    "data_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_000014.5_1</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000014.5_2</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_000014.5_3</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_000014.5_4</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_000014.5_5</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0    1\n",
       "0  NM_000014.5_1  A2M\n",
       "1  NM_000014.5_2  A2M\n",
       "2  NM_000014.5_3  A2M\n",
       "3  NM_000014.5_4  A2M\n",
       "4  NM_000014.5_5  A2M"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### First for exon-intron coordinates, we need both geneID and transcriptID\n",
    "data_transcriptsToGenes = pd.read_csv(\"../data/ExonIntronCoords_GeneNameTranscriptID_NCBI_RefSeq_hg38_FromGFFfile_5p.tsv\",sep=\"\\t\",header=None)\n",
    "print data_transcriptsToGenes.shape\n",
    "data_transcriptsToGenes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0_x    1_x      2              3  4            0_y      1_y\n",
       "0  chr1  12127  12327  NR_046018.2_1  +  NR_046018.2_1  DDX11L1\n",
       "1  chr1  12621  12821  NR_046018.2_2  +  NR_046018.2_2  DDX11L1\n",
       "2  chr1  14870  15070  NR_024540.1_1  -  NR_024540.1_1   WASH7P\n",
       "3  chr1  15696  15896  NR_024540.1_2  -  NR_024540.1_2   WASH7P\n",
       "4  chr1  16507  16707  NR_024540.1_3  -  NR_024540.1_3   WASH7P"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the two data sets\n",
    "data_all_5p = data_transcripts.merge(data_transcriptsToGenes,left_on=3,right_on=0)\n",
    "print data_all_5p.shape\n",
    "data_all_5p.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1_y</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       1_y   0_x    1_x      2  4\n",
       "0  DDX11L1  chr1  12127  12327  +\n",
       "1  DDX11L1  chr1  12621  12821  +\n",
       "2   WASH7P  chr1  14870  15070  -\n",
       "3   WASH7P  chr1  15696  15896  -\n",
       "4   WASH7P  chr1  16507  16707  -"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the SAF format\n",
    "data_all_5p_justGeneIDs = data_all_5p.iloc[:,[6,0,1,2,4]]\n",
    "print data_all_5p_justGeneIDs.shape\n",
    "data_all_5p_justGeneIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GeneID   Chr  Start    End Strand\n",
       "0  DDX11L1  chr1  12127  12327      +\n",
       "1  DDX11L1  chr1  12621  12821      +\n",
       "2   WASH7P  chr1  14870  15070      -\n",
       "3   WASH7P  chr1  15696  15896      -\n",
       "4   WASH7P  chr1  16507  16707      -"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_5p_justGeneIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_all_5p_justGeneIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all_5p_justGeneIDs.to_csv(\"../data/SAFformat_ExtractedExonIntronCoordinatesFromGFFfile_5p_OnlyGenes.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               3   0_x    1_x      2  4\n",
       "0  NR_046018.2_1  chr1  12127  12327  +\n",
       "1  NR_046018.2_2  chr1  12621  12821  +\n",
       "2  NR_024540.1_1  chr1  14870  15070  -\n",
       "3  NR_024540.1_2  chr1  15696  15896  -\n",
       "4  NR_024540.1_3  chr1  16507  16707  -"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the SAF format for transcriptID\n",
    "data_all_5p_justTranscriptIDs = data_all_5p.iloc[:,[3,0,1,2,4]]\n",
    "print data_all_5p_justTranscriptIDs.shape\n",
    "data_all_5p_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12127  12327      +\n",
       "1  NR_046018.2_2  chr1  12621  12821      +\n",
       "2  NR_024540.1_1  chr1  14870  15070      -\n",
       "3  NR_024540.1_2  chr1  15696  15896      -\n",
       "4  NR_024540.1_3  chr1  16507  16707      -"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_5p_justTranscriptIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_all_5p_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all_5p_justTranscriptIDs.to_csv(\"../data/SAFformat_ExtractedExonIntronCoordinatesFromGFFfile_5p_OnlyTranscriptIDs.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12513  12713  NR_046018.2_1  +\n",
       "1  chr1  13121  13321  NR_046018.2_2  +\n",
       "2  chr1  14729  14929  NR_024540.1_1  -\n",
       "3  chr1  14938  15138  NR_024540.1_2  -\n",
       "4  chr1  15847  16047  NR_024540.1_3  -"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### First for exon-intron coordinates, we need both geneID and transcriptID\n",
    "data_transcripts = pd.read_csv(\"../data/ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence_Just3p.bed\",sep=\"\\t\",header=None)\n",
    "print data_transcripts.shape\n",
    "data_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_000014.5_1</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000014.5_2</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_000014.5_3</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_000014.5_4</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_000014.5_5</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0    1\n",
       "0  NM_000014.5_1  A2M\n",
       "1  NM_000014.5_2  A2M\n",
       "2  NM_000014.5_3  A2M\n",
       "3  NM_000014.5_4  A2M\n",
       "4  NM_000014.5_5  A2M"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### First for exon-intron coordinates, we need both geneID and transcriptID\n",
    "data_transcriptsToGenes = pd.read_csv(\"../data/ExonIntronCoords_GeneNameTranscriptID_NCBI_RefSeq_hg38_FromGFFfile_3p.tsv\",sep=\"\\t\",header=None)\n",
    "print data_transcriptsToGenes.shape\n",
    "data_transcriptsToGenes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>0_y</th>\n",
       "      <th>1_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0_x    1_x      2              3  4            0_y      1_y\n",
       "0  chr1  12513  12713  NR_046018.2_1  +  NR_046018.2_1  DDX11L1\n",
       "1  chr1  13121  13321  NR_046018.2_2  +  NR_046018.2_2  DDX11L1\n",
       "2  chr1  14729  14929  NR_024540.1_1  -  NR_024540.1_1   WASH7P\n",
       "3  chr1  14938  15138  NR_024540.1_2  -  NR_024540.1_2   WASH7P\n",
       "4  chr1  15847  16047  NR_024540.1_3  -  NR_024540.1_3   WASH7P"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge the two data sets\n",
    "data_all_3p = data_transcripts.merge(data_transcriptsToGenes,left_on=3,right_on=0)\n",
    "print data_all_3p.shape\n",
    "data_all_3p.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1_y</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       1_y   0_x    1_x      2  4\n",
       "0  DDX11L1  chr1  12513  12713  +\n",
       "1  DDX11L1  chr1  13121  13321  +\n",
       "2   WASH7P  chr1  14729  14929  -\n",
       "3   WASH7P  chr1  14938  15138  -\n",
       "4   WASH7P  chr1  15847  16047  -"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the SAF format\n",
    "data_all_3p_justGeneIDs = data_all_3p.iloc[:,[6,0,1,2,4]]\n",
    "print data_all_3p_justGeneIDs.shape\n",
    "data_all_3p_justGeneIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    GeneID   Chr  Start    End Strand\n",
       "0  DDX11L1  chr1  12513  12713      +\n",
       "1  DDX11L1  chr1  13121  13321      +\n",
       "2   WASH7P  chr1  14729  14929      -\n",
       "3   WASH7P  chr1  14938  15138      -\n",
       "4   WASH7P  chr1  15847  16047      -"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_3p_justGeneIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_all_3p_justGeneIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all_3p_justGeneIDs.to_csv(\"../data/SAFformat_ExtractedExonIntronCoordinatesFromGFFfile_3p_OnlyGenes.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3</th>\n",
       "      <th>0_x</th>\n",
       "      <th>1_x</th>\n",
       "      <th>2</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               3   0_x    1_x      2  4\n",
       "0  NR_046018.2_1  chr1  12513  12713  +\n",
       "1  NR_046018.2_2  chr1  13121  13321  +\n",
       "2  NR_024540.1_1  chr1  14729  14929  -\n",
       "3  NR_024540.1_2  chr1  14938  15138  -\n",
       "4  NR_024540.1_3  chr1  15847  16047  -"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the SAF format\n",
    "data_all_3p_justTranscriptIDs = data_all_3p.iloc[:,[3,0,1,2,4]]\n",
    "print data_all_3p_justTranscriptIDs.shape\n",
    "data_all_3p_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12513  12713      +\n",
       "1  NR_046018.2_2  chr1  13121  13321      +\n",
       "2  NR_024540.1_1  chr1  14729  14929      -\n",
       "3  NR_024540.1_2  chr1  14938  15138      -\n",
       "4  NR_024540.1_3  chr1  15847  16047      -"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_3p_justTranscriptIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_all_3p_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all_3p_justTranscriptIDs.to_csv(\"../data/SAFformat_ExtractedExonIntronCoordinatesFromGFFfile_3p_OnlyTranscriptIDs.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's get exon-exon coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>15070</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15896</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16707</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12127  12327  NR_046018.2_1  +\n",
       "1  chr1  12621  12821  NR_046018.2_2  +\n",
       "2  chr1  14870  15070  NR_024540.1_1  -\n",
       "3  chr1  15696  15896  NR_024540.1_2  -\n",
       "4  chr1  16507  16707  NR_024540.1_3  -"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### First for exon-intron coordinates, we need both geneID and transcriptID\n",
    "data_transcripts = pd.read_csv(\"../data/ExtractedExonIntronCoordinates_5p_FromGFFfile_ForSequence_Just5p.bed\",sep=\"\\t\",header=None)\n",
    "print data_transcripts.shape\n",
    "data_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate into plus and minus\n",
    "data_transcripts_plus = data_transcripts[data_transcripts[4]==\"+\"]\n",
    "data_transcripts_minus = data_transcripts[data_transcripts[4]==\"-\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12127</td>\n",
       "      <td>12177</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12621</td>\n",
       "      <td>12671</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29939</td>\n",
       "      <td>29989</td>\n",
       "      <td>XR_001737835.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30567</td>\n",
       "      <td>30617</td>\n",
       "      <td>XR_001737835.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182646</td>\n",
       "      <td>182696</td>\n",
       "      <td>NR_148357.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end      transcriptID strand\n",
       "0   chr1   12127   12177     NR_046018.2_1      +\n",
       "1   chr1   12621   12671     NR_046018.2_2      +\n",
       "12  chr1   29939   29989  XR_001737835.1_1      +\n",
       "13  chr1   30567   30617  XR_001737835.1_2      +\n",
       "75  chr1  182646  182696     NR_148357.1_1      +"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_transcripts_exonexon = pd.DataFrame({\"chrom\":pd.concat([data_transcripts_plus[0],data_transcripts_minus[0]]),\n",
    "                                         \"start\":pd.concat([data_transcripts_plus[1],data_transcripts_minus[1]+50]),\n",
    "                                          \"end\":pd.concat([data_transcripts_plus[1]+50,data_transcripts_minus[2]]),\n",
    "                                          \"transcriptID\":pd.concat([data_transcripts_plus[3],data_transcripts_minus[3]]),\n",
    "                                          \"strand\":pd.concat([data_transcripts_plus[4],data_transcripts_minus[4]])},\n",
    "                                        columns=[\"chrom\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "data_transcripts_exonexon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12663</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13271</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30514</td>\n",
       "      <td>30614</td>\n",
       "      <td>XR_001737835.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30926</td>\n",
       "      <td>31026</td>\n",
       "      <td>XR_001737835.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183082</td>\n",
       "      <td>183182</td>\n",
       "      <td>NR_148357.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end      transcriptID strand\n",
       "0   chr1   12563   12663     NR_046018.2_1      +\n",
       "1   chr1   13171   13271     NR_046018.2_2      +\n",
       "12  chr1   30514   30614  XR_001737835.1_1      +\n",
       "13  chr1   30926   31026  XR_001737835.1_2      +\n",
       "76  chr1  183082  183182     NR_148357.1_1      +"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_transcripts_exonintron = pd.DataFrame({\"chrom\":pd.concat([data_transcripts_plus[0],data_transcripts_minus[0]]),\n",
    "                                         \"start\":pd.concat([data_transcripts_plus[1]+50,data_transcripts_minus[1]+50]),\n",
    "                                          \"end\":pd.concat([data_transcripts_plus[2]-50,data_transcripts_minus[2]-50]),\n",
    "                                          \"transcriptID\":pd.concat([data_transcripts_plus[3],data_transcripts_minus[3]]),\n",
    "                                          \"strand\":pd.concat([data_transcripts_plus[4],data_transcripts_minus[4]])},\n",
    "                                        columns=[\"chrom\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "data_transcripts_exonintron.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12713</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13321</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14779</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>14988</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>15897</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start    end   transcriptID strand\n",
       "0  chr1  12563  12713  NR_046018.2_1      +\n",
       "1  chr1  13171  13321  NR_046018.2_2      +\n",
       "2  chr1  14729  14779  NR_024540.1_1      -\n",
       "3  chr1  14938  14988  NR_024540.1_2      -\n",
       "4  chr1  15847  15897  NR_024540.1_3      -"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_transcripts_exonexon_sorted = data_transcripts_exonexon.sort_values(by=[\"chrom\",\"start\",\"end\"])\n",
    "data_transcripts_exonexon_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12663</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13271</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14779</td>\n",
       "      <td>14879</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14988</td>\n",
       "      <td>15088</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15897</td>\n",
       "      <td>15997</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start    end   transcriptID strand\n",
       "0  chr1  12563  12663  NR_046018.2_1      +\n",
       "1  chr1  13171  13271  NR_046018.2_2      +\n",
       "2  chr1  14779  14879  NR_024540.1_1      -\n",
       "3  chr1  14988  15088  NR_024540.1_2      -\n",
       "4  chr1  15897  15997  NR_024540.1_3      -"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_transcripts_exonintron_sorted = data_transcripts_exonintron.sort_values(by=[\"chrom\",\"start\",\"end\"])\n",
    "data_transcripts_exonintron_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12713</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13321</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14779</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>14988</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>15897</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12563  12713      +\n",
       "1  NR_046018.2_2  chr1  13171  13321      +\n",
       "2  NR_024540.1_1  chr1  14729  14779      -\n",
       "3  NR_024540.1_2  chr1  14938  14988      -\n",
       "4  NR_024540.1_3  chr1  15847  15897      -"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_5p_justTranscriptIDs = data_transcripts_exonexon_sorted.iloc[:,[3,0,1,2,4]]\n",
    "data_all_5p_justTranscriptIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_all_5p_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all_5p_justTranscriptIDs.to_csv(\"../data/SAFformat_ExtractedExonExonCoordinatesFromGFFfile_5p_OnlyTranscriptIDs.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12663</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13271</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14779</td>\n",
       "      <td>14879</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14988</td>\n",
       "      <td>15088</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15897</td>\n",
       "      <td>15997</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12563  12663      +\n",
       "1  NR_046018.2_2  chr1  13171  13271      +\n",
       "2  NR_024540.1_1  chr1  14779  14879      -\n",
       "3  NR_024540.1_2  chr1  14988  15088      -\n",
       "4  NR_024540.1_3  chr1  15897  15997      -"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_5p_justTranscriptIDs = data_transcripts_exonintron_sorted.iloc[:,[3,0,1,2,4]]\n",
    "data_all_5p_justTranscriptIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_all_5p_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all_5p_justTranscriptIDs.to_csv(\"../data/SAFformat_ExtractedIntronExonCoordinatesFromGFFfile_5p_OnlyTranscriptIDs-100bp.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1701899, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12713</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13321</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14929</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>15138</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>16047</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12513  12713  NR_046018.2_1  +\n",
       "1  chr1  13121  13321  NR_046018.2_2  +\n",
       "2  chr1  14729  14929  NR_024540.1_1  -\n",
       "3  chr1  14938  15138  NR_024540.1_2  -\n",
       "4  chr1  15847  16047  NR_024540.1_3  -"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### First for exon-intron coordinates, we need both geneID and transcriptID\n",
    "data_transcripts = pd.read_csv(\"../data/ExtractedExonIntronCoordinates_3p_FromGFFfile_ForSequence_Just3p.bed\",sep=\"\\t\",header=None)\n",
    "print data_transcripts.shape\n",
    "data_transcripts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate into plus and minus\n",
    "data_transcripts_plus = data_transcripts[data_transcripts[4]==\"+\"]\n",
    "data_transcripts_minus = data_transcripts[data_transcripts[4]==\"-\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12713</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13321</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30514</td>\n",
       "      <td>30664</td>\n",
       "      <td>XR_001737835.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30926</td>\n",
       "      <td>31076</td>\n",
       "      <td>XR_001737835.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183082</td>\n",
       "      <td>183232</td>\n",
       "      <td>NR_148357.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end      transcriptID strand\n",
       "0   chr1   12563   12713     NR_046018.2_1      +\n",
       "1   chr1   13171   13321     NR_046018.2_2      +\n",
       "12  chr1   30514   30664  XR_001737835.1_1      +\n",
       "13  chr1   30926   31076  XR_001737835.1_2      +\n",
       "76  chr1  183082  183232     NR_148357.1_1      +"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_transcripts_exonexon = pd.DataFrame({\"chrom\":pd.concat([data_transcripts_plus[0],data_transcripts_minus[0]]),\n",
    "                                         \"start\":pd.concat([data_transcripts_plus[1]+50,data_transcripts_minus[1]]),\n",
    "                                          \"end\":pd.concat([data_transcripts_plus[2],data_transcripts_minus[1]+50]),\n",
    "                                          \"transcriptID\":pd.concat([data_transcripts_plus[3],data_transcripts_minus[3]]),\n",
    "                                          \"strand\":pd.concat([data_transcripts_plus[4],data_transcripts_minus[4]])},\n",
    "                                        columns=[\"chrom\",\"start\",\"end\",\"transcriptID\",\"strand\"])\n",
    "data_transcripts_exonexon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12713</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13321</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14779</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>14988</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>15897</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start    end   transcriptID strand\n",
       "0  chr1  12563  12713  NR_046018.2_1      +\n",
       "1  chr1  13171  13321  NR_046018.2_2      +\n",
       "2  chr1  14729  14779  NR_024540.1_1      -\n",
       "3  chr1  14938  14988  NR_024540.1_2      -\n",
       "4  chr1  15847  15897  NR_024540.1_3      -"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_transcripts_exonexon_sorted = data_transcripts_exonexon.sort_values(by=[\"chrom\",\"start\",\"end\"])\n",
    "data_transcripts_exonexon_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12563</td>\n",
       "      <td>12713</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13171</td>\n",
       "      <td>13321</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14729</td>\n",
       "      <td>14779</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14938</td>\n",
       "      <td>14988</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15847</td>\n",
       "      <td>15897</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12563  12713      +\n",
       "1  NR_046018.2_2  chr1  13171  13321      +\n",
       "2  NR_024540.1_1  chr1  14729  14779      -\n",
       "3  NR_024540.1_2  chr1  14938  14988      -\n",
       "4  NR_024540.1_3  chr1  15847  15897      -"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_3p_justTranscriptIDs = data_transcripts_exonexon_sorted.iloc[:,[3,0,1,2,4]]\n",
    "data_all_3p_justTranscriptIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "data_all_3p_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_all_3p_justTranscriptIDs.to_csv(\"../data/SAFformat_ExtractedExonExonCoordinatesFromGFFfile_3p_OnlyTranscriptIDs.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get coordinates of introns that are less than 100 bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1316596, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12612</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13220</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14829</td>\n",
       "      <td>14969</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15038</td>\n",
       "      <td>15795</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15947</td>\n",
       "      <td>16606</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12227  12612  NR_046018.2_1  +\n",
       "1  chr1  12721  13220  NR_046018.2_2  +\n",
       "2  chr1  14829  14969  NR_024540.1_1  -\n",
       "3  chr1  15038  15795  NR_024540.1_2  -\n",
       "4  chr1  15947  16606  NR_024540.1_3  -"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intron_coords = pd.read_csv(\"../data/ExtractedIntronCoordinatesFromGFFfile_NoOverlapExon.bed\",sep=\"\\t\",header=None)\n",
    "print(intron_coords.shape)\n",
    "intron_coords.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(63714, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16765</td>\n",
       "      <td>16857</td>\n",
       "      <td>NR_024540.1_4</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>139696</td>\n",
       "      <td>139789</td>\n",
       "      <td>NR_039983.2_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chr1</td>\n",
       "      <td>187287</td>\n",
       "      <td>187379</td>\n",
       "      <td>XR_001737556.1_4</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>chr1</td>\n",
       "      <td>358183</td>\n",
       "      <td>358283</td>\n",
       "      <td>XR_002958507.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>chr1</td>\n",
       "      <td>494898</td>\n",
       "      <td>494991</td>\n",
       "      <td>NR_028322.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0       1       2                 3  4\n",
       "5   chr1   16765   16857     NR_024540.1_4  -\n",
       "23  chr1  139696  139789     NR_039983.2_1  -\n",
       "54  chr1  187287  187379  XR_001737556.1_4  -\n",
       "70  chr1  358183  358283  XR_002958507.1_2  -\n",
       "79  chr1  494898  494991     NR_028322.1_1  -"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get introns that are less than 100 bases\n",
    "introns_lessthan100 = intron_coords[intron_coords[2]-intron_coords[1]<=100]\n",
    "print(introns_lessthan100.shape)\n",
    "introns_lessthan100.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_lessthan100.columns= [\"chrom\",\"start\",\"end\",\"transcriptID\",\"strand\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1252882, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12612</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>13220</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14829</td>\n",
       "      <td>14969</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15038</td>\n",
       "      <td>15795</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15947</td>\n",
       "      <td>16606</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0      1      2              3  4\n",
       "0  chr1  12227  12612  NR_046018.2_1  +\n",
       "1  chr1  12721  13220  NR_046018.2_2  +\n",
       "2  chr1  14829  14969  NR_024540.1_1  -\n",
       "3  chr1  15038  15795  NR_024540.1_2  -\n",
       "4  chr1  15947  16606  NR_024540.1_3  -"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get introns that are greater than 100 bases\n",
    "introns_greaterthan100 = intron_coords[intron_coords[2]-intron_coords[1]>100]\n",
    "print(introns_greaterthan100.shape)\n",
    "introns_greaterthan100.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split up introns that are plus and minus strand\n",
    "introns_greaterthan100_plus = introns_greaterthan100[introns_greaterthan100[4]==\"+\"]\n",
    "introns_greaterthan100_minus = introns_greaterthan100[introns_greaterthan100[4]==\"-\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get 5' ends of plus and minus\n",
    "introns_greaterthan100_5p = pd.DataFrame({\"chrom\":pd.concat([introns_greaterthan100_plus[0],introns_greaterthan100_minus[0]]),\n",
    "                                         \"start\":pd.concat([introns_greaterthan100_plus[1],introns_greaterthan100_minus[2]-100]),\n",
    "                                          \"end\":pd.concat([introns_greaterthan100_plus[1]+100,introns_greaterthan100_minus[2]]),\n",
    "                                          \"transcriptID\":pd.concat([introns_greaterthan100_plus[3],introns_greaterthan100_minus[3]]),\n",
    "                                          \"strand\":pd.concat([introns_greaterthan100_plus[4],introns_greaterthan100_minus[4]])},\n",
    "                                        columns=[\"chrom\",\"start\",\"end\",\"transcriptID\",\"strand\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1252882, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30667</td>\n",
       "      <td>30767</td>\n",
       "      <td>XR_001737835.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>182846</td>\n",
       "      <td>NR_148357.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183240</td>\n",
       "      <td>183340</td>\n",
       "      <td>NR_148357.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end      transcriptID strand\n",
       "0   chr1   12227   12327     NR_046018.2_1      +\n",
       "1   chr1   12721   12821     NR_046018.2_2      +\n",
       "11  chr1   30667   30767  XR_001737835.1_2      +\n",
       "49  chr1  182746  182846     NR_148357.1_1      +\n",
       "50  chr1  183240  183340     NR_148357.1_2      +"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(introns_greaterthan100_5p.shape)\n",
    "introns_greaterthan100_5p.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1252882, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30667</td>\n",
       "      <td>30767</td>\n",
       "      <td>XR_001737835.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr1</td>\n",
       "      <td>182746</td>\n",
       "      <td>182846</td>\n",
       "      <td>NR_148357.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183240</td>\n",
       "      <td>183340</td>\n",
       "      <td>NR_148357.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end      transcriptID strand\n",
       "0   chr1   12227   12327     NR_046018.2_1      +\n",
       "1   chr1   12721   12821     NR_046018.2_2      +\n",
       "11  chr1   30667   30767  XR_001737835.1_2      +\n",
       "49  chr1  182746  182846     NR_148357.1_1      +\n",
       "50  chr1  183240  183340     NR_148357.1_2      +"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# concantenate the 5p with introns less than 100\n",
    "#introns_5p = pd.concat([introns_lessthan100,introns_greaterthan100_5p],axis=0)\n",
    "introns_5p = introns_greaterthan100_5p\n",
    "print(introns_5p.shape)\n",
    "introns_5p.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12327</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>12821</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14869</td>\n",
       "      <td>14969</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15695</td>\n",
       "      <td>15795</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16506</td>\n",
       "      <td>16606</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start    end   transcriptID strand\n",
       "0  chr1  12227  12327  NR_046018.2_1      +\n",
       "1  chr1  12721  12821  NR_046018.2_2      +\n",
       "2  chr1  14869  14969  NR_024540.1_1      -\n",
       "3  chr1  15695  15795  NR_024540.1_2      -\n",
       "4  chr1  16506  16606  NR_024540.1_3      -"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sort introns \n",
    "introns_5p_sorted = introns_5p.sort_values(by=[\"chrom\",\"start\",\"end\"])\n",
    "introns_5p_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12227</td>\n",
       "      <td>12327</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12721</td>\n",
       "      <td>12821</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14869</td>\n",
       "      <td>14969</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15695</td>\n",
       "      <td>15795</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16506</td>\n",
       "      <td>16606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12227  12327      +\n",
       "1  NR_046018.2_2  chr1  12721  12821      +\n",
       "2  NR_024540.1_1  chr1  14869  14969      -\n",
       "3  NR_024540.1_2  chr1  15695  15795      -\n",
       "4  NR_024540.1_3  chr1  16506  16606      -"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "introns_5p_sorted_justTranscriptIDs = introns_5p_sorted.iloc[:,[3,0,1,2,4]]\n",
    "introns_5p_sorted_justTranscriptIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "introns_5p_sorted_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_5p_sorted_justTranscriptIDs[\"Start\"] = introns_5p_sorted_justTranscriptIDs[\"Start\"] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12228</td>\n",
       "      <td>12327</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12722</td>\n",
       "      <td>12821</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14870</td>\n",
       "      <td>14969</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15696</td>\n",
       "      <td>15795</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16507</td>\n",
       "      <td>16606</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12228  12327      +\n",
       "1  NR_046018.2_2  chr1  12722  12821      +\n",
       "2  NR_024540.1_1  chr1  14870  14969      -\n",
       "3  NR_024540.1_2  chr1  15696  15795      -\n",
       "4  NR_024540.1_3  chr1  16507  16606      -"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "introns_5p_sorted_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_5p_sorted_justTranscriptIDs.to_csv(\"../data/SAFformat_ExtractedIntronCoordinatesFromGFFfile_5p_OnlyTranscriptIDs.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get 3' ends of plus and minus\n",
    "introns_greaterthan100_3p = pd.DataFrame({\"chrom\":pd.concat([introns_greaterthan100_plus[0],introns_greaterthan100_minus[0]]),\n",
    "                                         \"start\":pd.concat([introns_greaterthan100_plus[2]-100,introns_greaterthan100_minus[1]]),\n",
    "                                          \"end\":pd.concat([introns_greaterthan100_plus[2],introns_greaterthan100_minus[1]+100]),\n",
    "                                          \"transcriptID\":pd.concat([introns_greaterthan100_plus[3],introns_greaterthan100_minus[3]]),\n",
    "                                          \"strand\":pd.concat([introns_greaterthan100_plus[4],introns_greaterthan100_minus[4]])},\n",
    "                                        columns=[\"chrom\",\"start\",\"end\",\"transcriptID\",\"strand\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1252882, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12512</td>\n",
       "      <td>12612</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13120</td>\n",
       "      <td>13220</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30875</td>\n",
       "      <td>30975</td>\n",
       "      <td>XR_001737835.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183031</td>\n",
       "      <td>183131</td>\n",
       "      <td>NR_148357.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183639</td>\n",
       "      <td>183739</td>\n",
       "      <td>NR_148357.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end      transcriptID strand\n",
       "0   chr1   12512   12612     NR_046018.2_1      +\n",
       "1   chr1   13120   13220     NR_046018.2_2      +\n",
       "11  chr1   30875   30975  XR_001737835.1_2      +\n",
       "49  chr1  183031  183131     NR_148357.1_1      +\n",
       "50  chr1  183639  183739     NR_148357.1_2      +"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(introns_greaterthan100_3p.shape)\n",
    "introns_greaterthan100_3p.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1252882, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12512</td>\n",
       "      <td>12612</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13120</td>\n",
       "      <td>13220</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>30875</td>\n",
       "      <td>30975</td>\n",
       "      <td>XR_001737835.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183031</td>\n",
       "      <td>183131</td>\n",
       "      <td>NR_148357.1_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183639</td>\n",
       "      <td>183739</td>\n",
       "      <td>NR_148357.1_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end      transcriptID strand\n",
       "0   chr1   12512   12612     NR_046018.2_1      +\n",
       "1   chr1   13120   13220     NR_046018.2_2      +\n",
       "11  chr1   30875   30975  XR_001737835.1_2      +\n",
       "49  chr1  183031  183131     NR_148357.1_1      +\n",
       "50  chr1  183639  183739     NR_148357.1_2      +"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# concantenate the 3p with introns less than 100\n",
    "#introns_3p = pd.concat([introns_lessthan100,introns_greaterthan100_3p],axis=0)\n",
    "introns_3p = introns_greaterthan100_3p\n",
    "print(introns_3p.shape)\n",
    "introns_3p.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>transcriptID</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12512</td>\n",
       "      <td>12612</td>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13120</td>\n",
       "      <td>13220</td>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14829</td>\n",
       "      <td>14929</td>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15038</td>\n",
       "      <td>15138</td>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15947</td>\n",
       "      <td>16047</td>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start    end   transcriptID strand\n",
       "0  chr1  12512  12612  NR_046018.2_1      +\n",
       "1  chr1  13120  13220  NR_046018.2_2      +\n",
       "2  chr1  14829  14929  NR_024540.1_1      -\n",
       "3  chr1  15038  15138  NR_024540.1_2      -\n",
       "4  chr1  15947  16047  NR_024540.1_3      -"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sort introns \n",
    "introns_3p_sorted = introns_3p.sort_values(by=[\"chrom\",\"start\",\"end\"])\n",
    "introns_3p_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12512</td>\n",
       "      <td>12612</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13120</td>\n",
       "      <td>13220</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14829</td>\n",
       "      <td>14929</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15038</td>\n",
       "      <td>15138</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15947</td>\n",
       "      <td>16047</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12512  12612      +\n",
       "1  NR_046018.2_2  chr1  13120  13220      +\n",
       "2  NR_024540.1_1  chr1  14829  14929      -\n",
       "3  NR_024540.1_2  chr1  15038  15138      -\n",
       "4  NR_024540.1_3  chr1  15947  16047      -"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "introns_3p_sorted_justTranscriptIDs = introns_3p_sorted.iloc[:,[3,0,1,2,4]]\n",
    "introns_3p_sorted_justTranscriptIDs.columns = [\"GeneID\",\"Chr\",\"Start\",\"End\",\"Strand\"]\n",
    "introns_3p_sorted_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NR_046018.2_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12513</td>\n",
       "      <td>12612</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NR_046018.2_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13121</td>\n",
       "      <td>13220</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NR_024540.1_1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14830</td>\n",
       "      <td>14929</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NR_024540.1_2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15039</td>\n",
       "      <td>15138</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NR_024540.1_3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>15948</td>\n",
       "      <td>16047</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID   Chr  Start    End Strand\n",
       "0  NR_046018.2_1  chr1  12513  12612      +\n",
       "1  NR_046018.2_2  chr1  13121  13220      +\n",
       "2  NR_024540.1_1  chr1  14830  14929      -\n",
       "3  NR_024540.1_2  chr1  15039  15138      -\n",
       "4  NR_024540.1_3  chr1  15948  16047      -"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "introns_3p_sorted_justTranscriptIDs[\"Start\"] = introns_3p_sorted_justTranscriptIDs[\"Start\"] + 1\n",
    "introns_3p_sorted_justTranscriptIDs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "introns_3p_sorted_justTranscriptIDs.to_csv(\"../data/SAFformat_ExtractedIntronCoordinatesFromGFFfile_3p_OnlyTranscriptIDs.bed\",sep=\"\\t\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
